monitor/event_stats.hpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2015+ Budnik Andrey <budnik27@gmail.com>
+ *
+ * This file is part of Elliptics.
+ *
+ * Elliptics is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Elliptics is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Elliptics.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef EVENT_STATS_HPP
+#define EVENT_STATS_HPP
+
+#include "cache/treap.hpp"
+
+#include <mutex>
+#include <algorithm>
+#include <functional> // not2
+#include <vector>
+#include <stack>
+
+namespace ioremap { namespace monitor {
+
+class mutex_lock_policy
+{
+	typedef std::mutex mutex_t;
+public:
+	class unique_lock {
+	public:
+		unique_lock(mutex_lock_policy *policy)
+		: lock_(policy->get_lock())
+		{}
+	private:
+		std::unique_lock<mutex_t> lock_;
+	};
+
+	mutex_t &get_lock() { return mut_; }
+private:
+	mutex_t mut_;
+};
+
+class null_lock_policy
+{
+public:
+	class unique_lock {
+	public:
+		unique_lock(null_lock_policy *policy) { (void) policy; }
+	};
+};
```

Comment:
`template<typename E, typename lock_policy = null_lock_policy>
class event_stats : private lock_policy {...};`

event_stats used in the same way as stl containers used. But stl containers doesn't provide native thread-safety.
It is convinient to use thread-safety policy as an event_stats class template parameter.
Tests use null_lock_policy, and top_provider use mutex_lock_policy